crates/noirc_frontend/src/hir/def_collector/dc_crate.rs

```diff
@@ -341,6 +478,10 @@ fn resolve_functions(
             let mut resolver = Resolver::new(interner, &path_resolver, def_maps, file_id);
             resolver.set_self_type(self_type);
 
+            let mut resolved_const_ids =
+                resolve_global_constants(&mut resolver, collected_consts.clone());
+            file_const_ids.append(&mut resolved_const_ids);
```

Comment:
Your assumption that I am re-resolving is correct. Each function uses its own resolver (which has its own scope object) with how the current resolver infra is set up. Thus, it is necessary to add a global const to a given functions scopes using that funcs resolver (without a bigger refactor). 

I was going to test further to make sure conflicting variable declared within an inner scope of a function are checked. The current functionality is as such: for `find_variable` inside the resolver I check for global variables first, and if they exist move on. Thus, if an internal function variable of the same name exists it will be excluded and error out that it is unused. I don't really like this functionality. I tested out how Rust does things in a REPL, and it looks like Rust just uses the variable of lower scope first, and ignores the variable in the global scope, allowing it to be unused. We could follow this functionality. Or move to error'ing out on a conflicting internal and global var. I'm inclined to follow the same methodology as Rust.

Also, I did change to only interning each constant during `collect_global_constants`, while it gets re-resolved for each function without interning the const stmt.

Reply:
I changed the find_variable method to operate like Rust where it looks for the variable in the current scope first. If it doesn't exist then we will look in the resolver's global scope to see if we have a global const defined. Global consts are currently allowed to remain unused, however, this can be changed and we require global consts to be used as well.